File: play/src/engine/playData/skin.ts
import { SkinSpriteName } from "@sonolus/core";

export const skin = defineSkin({
  renderMode: "default",
  sprites: {

    //background
    backgroundNormal: "Background Normal",
    backgroundChallenge: "Background Challenge",


    //note
    noteTapNormal: "Tap Note Normal",
    noteTapChallenge: "Tap Note Challenge",

    noteDrag: "Drag Note",

    noteHold: "Hold Note",
    noteHoldStartNormal: "Hold Start Normal",
    noteHoldStartChallenge: "Hold Start Challenge",
    noteHoldConnectorNormal: "Hold Connector Normal",
    noteHoldConnectorChallenge: "Hold Connector Challenge",

    //lines
    line0: "L0",
    ...Object.fromEntries(
      Array.from({ length: 31 /* number of lines available in skin*/ }, (_, i) => [`line${i + 1}`, `L${i + 1}`])
    ),

    //JudgeRings
    judgeRing0: "J0",
    ...Object.fromEntries(
      Array.from({ length: 31 }, (_, i) => [`judgeRing${i + 1}`, `J${i + 1}`])
    )
  },
})

File: pyline/lib/skin.py
from typing import Any

from sonolus.script.sprite import (
    RenderMode,
    Sprite,
    SpriteGroup,
    skin,
    sprite,
    sprite_group,
)


def colored_sprites(name: str, count: int) -> Any:
    return sprite_group(f"{name} Color {v}" for v in range(count))


@skin
class Skin:
    render_mode: RenderMode = RenderMode.LIGHTWEIGHT
    # Drawing order: background < lines < hold connectors < notes < judge rings

    background_half_disc: SpriteGroup = colored_sprites("Background Half Disc", 18)
    background_fade: SpriteGroup = colored_sprites("Fade Out", 18)

    # notes
    note_drag: Sprite = sprite("Drag Note")
    note_hold: Sprite = sprite("Hold Note")

    note_tap: SpriteGroup = colored_sprites("Tap Note", 18)
    note_hold_head_decorator: Sprite = sprite("Hold Head Decorator")
    note_hold_connector: SpriteGroup = colored_sprites("Hold Connector", 18)
    note_hold_tail: SpriteGroup = colored_sprites("Hold Connector Fade Out", 18)
    miss_effect: Sprite = sprite("Miss Effect Overlay")
    miss_cross: Sprite = sprite("Miss Effect Cross")

    line_discs: SpriteGroup = colored_sprites("Line Disc", 64)
    pixel: SpriteGroup = colored_sprites("Pixel", 64)
    """Pixel is used for lines, background, ui"""

    judge_rings: SpriteGroup = colored_sprites("Judge Ring", 32)
